lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,324 @@
+        # when coupled with the handle property.
+        super(_HKEY_CONSTANT, self).__init__(hkey_name, None)
+
+    def _get_hkey(self, key):
+        return getattr(winreg, key)
+
+    @property
+    def handle(self):
+        if not self._handle:
+            self._handle = self._get_hkey(self.path)
+        return self._handle
+
+    @handle.setter
+    def handle(self, handle):
+        if not self._handle:
+            self._handle = handle
+
+
+class HKEY(object):
+    """
+    Predefined, open registry HKEYs
+    From the Microsoft docs:
+    An application must open a key before it can read data from the registry.
+    To open a key, an application must supply a handle to another key in
+    the registry that is already open. The system defines predefined keys
+    that are always open. Predefined keys help an application navigate in
+    the registry."""
+
+    HKEY_CLASSES_ROOT = _HKEY_CONSTANT("HKEY_CLASSES_ROOT")
+    HKEY_CURRENT_USER = _HKEY_CONSTANT("HKEY_CURRENT_USER")
+    HKEY_USERS = _HKEY_CONSTANT("HKEY_USERS")
+    HKEY_LOCAL_MACHINE = _HKEY_CONSTANT("HKEY_LOCAL_MACHINE")
+    HKEY_CURRENT_CONFIG = _HKEY_CONSTANT("HKEY_CURRENT_CONFIG")
+    HKEY_PERFORMANCE_DATA = _HKEY_CONSTANT("HKEY_PERFORMANCE_DATA")
+
+
+class WindowsRegistryView(object):
+    """
+    Interface to provide access, querying, and searching to Windows registry entries.
+    This class represents a single key entrypoint into the Windows registry
+    and provides an interface to this key's values, its subkeys, and those subkey's values.
+    This class cannot be used to move freely about the registry, only subkeys/values of
+    the root key used to instantiate this class.
+    """
+
+    def __init__(self, key, root_key=HKEY.HKEY_CURRENT_USER):
+        """Constructs a Windows Registry entrypoint to key provided
+        root_key should be an already open root key or an hkey constant if provided
+
+        Args:
+            key (str): registry key to provide root for registry key for this clas
+            root_key: Already open registry key or HKEY constant to provide access into
+                         the Windows registry. Registry access requires an already open key
+                         to get an entrypoint, the HKEY constants are always open, or an already
+                         open key can be used instead.
+        """
+        if not is_windows:
+            raise RuntimeError(
+                "Cannot instantiate Windows Registry class on non Windows platforms"
+            )
+        self.key = key
+        self.root = root_key
+        self._reg = None
+
+    @contextmanager
+    def invalid_reg_ref_error_handler(self):
+        try:
+            try:
+                yield
+            except FileNotFoundError as e:
+                if e.winerror == 2:
+                    tty.debug("Key %s at position %s does not exist" % (self.key, str(self.root)))
+                else:
+                    raise e
+        finally:
```

Comment:
Unsure, I think this may have been a rebasing mistake, or maybe I got turned around during implementation, either way, the nesting is unnecessary. Removed.